src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/14.diff.sql

```diff
@@ -0,0 +1,537 @@
+     NOT EXISTS (SELECT 1 FROM dbo.ResourceChangeType WHERE ResourceChangeTypeId = 0))
+BEGIN
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (0, N'Creation')
+END
+GO
+
+IF (EXISTS (SELECT 1 FROM sys.tables WHERE name = 'ResourceChangeType') AND
+     NOT EXISTS (SELECT 1 FROM dbo.ResourceChangeType WHERE ResourceChangeTypeId = 1))
+BEGIN
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (1, N'Update')
+END
+GO
+
+IF (EXISTS (SELECT 1 FROM sys.tables WHERE name = 'ResourceChangeType') AND
+     NOT EXISTS (SELECT 1 FROM dbo.ResourceChangeType WHERE ResourceChangeTypeId = 2))
+BEGIN
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (2, N'Deletion')
+END
+GO
+
+/*************************************************************
+    Stored procedures for capturing and fetching resource changes
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     CaptureResourceChanges
+--
+-- DESCRIPTION
+--     Inserts resource change data
+--
+-- PARAMETERS
+--     @isDeleted
+--         * Whether this resource marks the resource as deleted.
+--     @version
+--         * The version of the resource being written
+--     @resourceId
+--         * The resource ID
+--     @resourceTypeId
+--         * The ID of the resource type
+--
+-- RETURN VALUE
+--     It does not return a value.
+--
+CREATE OR ALTER PROCEDURE dbo.CaptureResourceChanges
+    @isDeleted bit, 
+    @version int, 
+    @resourceId varchar(64),
+    @resourceTypeId smallint
+AS
+BEGIN
+    DECLARE @changeType SMALLINT
+    IF (@isDeleted = 1) BEGIN
+        SET @changeType = 2    /* DELETION */
+    END
+    ELSE BEGIN
+        IF (@version = 1) BEGIN
+            SET @changeType = 0 /* CREATION */
+        END
+        ELSE BEGIN
+            SET @changeType = 1 /* UPDATE */
+        END
+    END
+
+    INSERT INTO dbo.ResourceChangeData
+        (ResourceId, ResourceTypeId, ResourceVersion, ResourceChangeTypeId)
+    VALUES
+        (@resourceId, @resourceTypeId, @version, @changeType)
+END
+GO
+
+--
+-- STORED PROCEDURE
+--     FetchResourceChanges
+--
+-- DESCRIPTION
+--     Returns the number of resource change records from startId. The start id is inclusive.
+--
+-- PARAMETERS
+--     @startId
+--         * The start id of resource change records to fetch.
+--     @pageSize
+--         * The page size for fetching resource change records.
+--
+-- RETURN VALUE
+--     Resource change data rows.
+--
+CREATE OR ALTER PROCEDURE dbo.FetchResourceChanges
+    @startId bigint, 
+    @pageSize int
+AS
+BEGIN
+
+    SET NOCOUNT ON;
+
+    -- Given the fact that Read Committed Snapshot isolation level is enabled on the FHIR database, 
+    -- using the Repeatable Read isolation level to avoid skipping resource changes 
+    -- due to interleaved transactions on the resource change data table.
+    SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
```

Comment:
So that I understand, would using the `UPDLOCK` hint also work in this scenario? If so, why is it preferable to instead change the isolation level?

Reply:
Good point. Thanks. Changed it to a Repeatable Read table hint instead of setting a Repeatable Read isolation level. Based on my research, using a repeatable read table hint is more suitable for our scenario due to the Updlock table hint acquires an [update (U) lock](https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2008-r2/ms175519(v=sql.105)#update-locks) on the resource which avoids a potential deadlock problem but blocks other read transactions for the same resource until the transaction completes. On the other hand, the Repeatable Read hint acquires a [shared (S) lock](https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2008-r2/ms175519(v=sql.105)#shared-locks) which allows other transactions to read the same resource in concurrent transactions. In our case, we insert and read often, don't update the resource change data, and don't delete the resource at the same as reading the resource. So,  I think that shared (S) locks on a resource are appropriate for this scenario.